### PR TITLE
Packages: Import eslint packages

### PR DESCRIPTION
**Pull request by sirreal**
#### Changes proposed in this Pull Request

* Modify build packages script to skip build if `src` dir is absent from a package
* Pull in `eslint-config-wpcalypso`
* Pull in `eslint-plugin-wpcalypso`
* Update to use Calypso versions of eslint and babel-eslint
* Update tests to use Jest
* Eslint fixes

I had to cherry pick b77a6ce6731e8ddf3539e628ff648ac78c7b63a5 from #29260 because we were parsing `node_modules` in packages and the parser couldn't parse some things.

#### Testing instructions

* Eslint should continue to work, config and plugins from these modules should work as expected.
* `test-packages` job on CI should run the test suite from `eslint-plugin-wpcalypso`

### CODE REVIEW

**Comment by sirreal** [comment]
I've rebased this and squashed the prepare changes into a single commit. Ready for review.

**Comment by sirreal** [comment]
Needs update-deps before review… hold off.

**Comment by sirreal** [comment]
This should be ready now 👍

File: packages/eslint-config-wpcalypso/.eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/** @format */
+
+module.exports = {
+	extends: '../../.eslintrc.js',
```
Comment (sirreal) [inline packages/eslint-config-wpcalypso/.eslintrc.js]: These can actually be removed, [eslint will look in parent directories until a the root dir or a `root: true` config.](https://eslint.org/docs/user-guide/configuring#using-configuration-files-1)

File: packages/eslint-plugin-wpcalypso/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/** @format */
+
+module.exports = {
+	extends: '../../.eslintrc.js',
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/.eslintrc.js]: Drop this

File: packages/eslint-plugin-wpcalypso/LICENSE.md [hunk 3]
```diff
@@ -0,0 +1,264 @@
+The GNU General Public License, Version 2, June 1991 (GPLv2)
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/LICENSE.md]: Drop file

File: packages/eslint-plugin-wpcalypso/README.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+{
+    "rules": {
+        "wpcalypso/rule-name": 2
+    }
+}
+```
+
+## Supported Rules
+
+- [`i18n-ellipsis`](docs/rules/i18n-ellipsis.md): Disallow using three dots in translate strings
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/README.md]: Check whether these paths are OK

File: packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+		return this.translate( 'Hello World' );
+	}
+} );
+```
+
+The following patterns are not warnings:
+
+```js
+import i18n from 'i18n-calypso';
+i18n.translate('Hello World');
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md]: spaces!

File: packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# Disallow this.translate()
+Disallows the usage of `this.translate()` in favor of the high order component `localize()` from `i18n-calypso` repository.
+
+## Rule Details
+
+This rules aims to prevent the usage of `this.translate` that uses the mixin approach which makes the components harder to test.
+The following is considered a warning:
+```js
+const MyComponent = React.createClass( {
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md]: Drop `createClass`

File: packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	render() {
+		return this.translate( 'Hello World' );
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md]: `this.props.translate`

File: packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+The following patterns are not warnings:
+
+```js
+import i18n from 'i18n-calypso';
+i18n.translate('Hello World');
+```
+
+```js
+import { localize } from 'i18n-calypso';
+const MyComponent = React.createClass( {
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md]: Fix class

File: packages/eslint-plugin-wpcalypso/docs/rules/jsx-classname-namespace.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+/*eslint jsx-classname-namespace: [ "error", { rootFiles: [ 'foo.js' ] } ]*/
+// client/sample-component/foo.js
+export default function() {
+	return <div className="sample" />;
+}
+```
+
+Examples of **incorrect** code for this rule with `rootFiles` set to `[ 'foo.js' ]`: (watch the filename)
+
+```js
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/docs/rules/jsx-classname-namespace.md]: jsx

File: packages/eslint-plugin-wpcalypso/docs/rules/jsx-classname-namespace.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+		{
+			"rootFiles": [ "index.jsx", "main.jsx" ]
+		}
+	]
+}
+```
+
+Examples of **correct** code for this rule with `rootFiles` set to `[ 'foo.js' ]`: (watch the filename)
+
+```js
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/docs/rules/jsx-classname-namespace.md]: jsx

File: packages/eslint-plugin-wpcalypso/docs/rules/jsx-classname-namespace.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+		<div className="another-sample">
+			<div className="another-sample-child" />
+		</div>
+	);
+}
+```
+
+The following patterns are not warnings:
+
+```js
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/docs/rules/jsx-classname-namespace.md]: jsx

File: packages/eslint-plugin-wpcalypso/docs/rules/jsx-classname-namespace.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+If JSX element includes `className`, it must include the containing directory name as the component name for the root element, or as a component prefix for child elements (with `__` separating component and element names).
+
+[See Calypso CSS guidelines for more information](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md)
+
+## Rule Details
+
+The following patterns are considered warnings:
+
+```js
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/docs/rules/jsx-classname-namespace.md]: jsx

File: packages/eslint-plugin-wpcalypso/docs/rules/jsx-gridicon-size.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+Gridicon JSX elements must use one of the recommended sizes.
+
+In previous incarnations, this warning could be subdued by adding a `nonStandardSize` prop to the element, but it's recommended instead to disable the rule using standard ESLint rule options ([documentation](http://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments)).
+
+## Rule Details
+
+The following patterns are considered warnings:
+
+```js
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/docs/rules/jsx-gridicon-size.md]: jsx here and below

File: packages/eslint-plugin-wpcalypso/lib/rules/index.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
+/** @format */
+
+module.exports = {
```
Comment (sirreal) [inline packages/eslint-plugin-wpcalypso/lib/rules/index.js]: Just lift this to the main index

**Review by sirreal — COMMENTED** [review]
Some notes.

This has not been changed to es6, namely it doesn't use `import/export`. Doesn't really seem worth it.

File: packages/eslint-plugin-wpcalypso/README.md (hunk 4 above)
Comment (sirreal) [reply packages/eslint-plugin-wpcalypso/README.md]: They are 👍

**Review by sirreal — COMMENTED** [review]
Rebased with small fixes and a frew update-deps. Ready for review.

File: packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md (hunk 7 above)
Comment (sirreal) [reply packages/eslint-plugin-wpcalypso/docs/rules/i18n-no-this-translate.md]: The `createClass` is already gone. The `this.translate` is showing a lint error from the mixin days.

**Review by jsnajdr — APPROVED** [review]
Looks ready to merge :+1:

There's one change I think we should do ASAP once the ESLint packages are in our monorepo: unify `eslint-config-wpcalypso` and `eslint-plugin-wpcalypso` into one package. ESLint plugins [can export both rules and configs](https://github.com/yannickcr/eslint-plugin-react#configuration). This feature probably wasn't available back when our ESLint packages were created.

See, for example, how [Jest does it](https://github.com/jest-community/eslint-plugin-jest/blob/master/index.js). The object exported from the plugin module has keys `configs` and `rules`.

**Comment by sirreal** [comment]
Thanks for review.

> There's one change I think we should do ASAP once the ESLint packages are in our monorepo

Sounds like a good plan! We might pull them both into `@automattic/eslint-calypso` (or similar). I'd also like to review our overrides and pull anything that makes sense into the package and release a v1 of the new package.

**Comment by jsnajdr** [comment]
> I'd also like to review our overrides and pull anything that makes sense into the package and release a v1 of the new package.

Yes! For example, Calypso `.eslintrc` disables a few `jsx-a11y` rules that don't work well in our codebase. And extends the `prettier` ESLint plugin to disable the code-formatting rules. When experimenting with ESLint upgrade in Jetpack repo (FYI @brbrr), I found out they need exactly the same overrides as Calypso. Can be part of the common `eslint-plugin-calypso` config.

**Pull request merged**